File: pkg/tanka/export_test.go
package tanka

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"github.com/grafana/tanka/pkg/jsonnet"
	"github.com/grafana/tanka/pkg/kubernetes/manifest"
	"github.com/rs/zerolog"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"k8s.io/apimachinery/pkg/labels"
)

func Test_replaceTmplText(t *testing.T) {
	type args struct {
		s   string
		old string
		new string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		{"text only", args{"a", "a", "b"}, "b"},
		{"action blocks", args{"{{a}}{{.}}", "a", "b"}, "{{a}}{{.}}"},
		{"mixed", args{"a{{a}}a{{a}}a", "a", "b"}, "b{{a}}b{{a}}b"},
		{"invalid template format handled as text", args{"a}}a{{a", "a", "b"}, "b}}b{{b"},
		{
			name: "keep path separator in action block",
			args: args{`{{index .metadata.labels "app.kubernetes.io/name"}}/{{.metadata.name}}`, "/", BelRune},
			want: "{{index .metadata.labels \"app.kubernetes.io/name\"}}\u0007{{.metadata.name}}",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := replaceTmplText(tt.args.s, tt.args.old, tt.args.new); got != tt.want {
				t.Errorf("replaceInTmplText() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestExportEnvironments(t *testing.T) {
	tempDir := t.TempDir()
	require.NoError(t, os.Chdir("testdata"))
	defer func() { require.NoError(t, os.Chdir("..")) }()

	// Find envs
	envs, err := FindEnvs(t.Context(), "test-export-envs", FindOpts{Selector: labels.Everything()})
	require.NoError(t, err)

	// Export all envs
	opts := &ExportEnvOpts{
		Format:    "{{.metadata.namespace}}/{{.metadata.name}}",
		Extension: "yaml",
	}
	opts.Opts.ExtCode = jsonnet.InjectedCode{
		"deploymentName": "'initial-deployment'",
		"serviceName":    "'initial-service'",
	}
	require.NoError(t, ExportEnvironments(t.Context(), envs, tempDir, opts))
	checkFiles(t, tempDir, []string{
		filepath.Join(tempDir, "inline-namespace1", "my-configmap.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-service.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-service.yaml"),
		filepath.Join(tempDir, "static", "initial-deployment.yaml"),
		filepath.Join(tempDir, "static", "initial-service.yaml"),
		filepath.Join(tempDir, "manifest.json"),
	})
	manifestContent, err := os.ReadFile(filepath.Join(tempDir, "manifest.json"))
	require.NoError(t, err)
	assert.Equal(t, string(manifestContent), `{
    "inline-namespace1/my-configmap.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace1/my-deployment.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace1/my-service.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace2/my-deployment.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace2/my-service.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "static/initial-deployment.yaml": "test-export-envs/static-env/main.jsonnet",
    "static/initial-service.yaml": "test-export-envs/static-env/main.jsonnet"
}`)

	// Try to re-export
	assert.EqualError(t, ExportEnvironments(t.Context(), envs, tempDir, opts), fmt.Sprintf("output dir `%s` not empty. Pass a different --merge-strategy to ignore this", tempDir))

	// Try to re-export with the --merge-strategy=fail-on-conflicts flag. Will still fail because Tanka will not overwrite manifests silently
	opts.MergeStrategy = ExportMergeStrategyFailConflicts
	assert.ErrorContains(t, ExportEnvironments(t.Context(), envs, tempDir, opts), "already exists. Aborting")

	// Re-export only one env with --merge-stategy=replace-envs flag
	opts.Opts.ExtCode = jsonnet.InjectedCode{
		"deploymentName": "'updated-deployment'",
		"serviceName":    "'updated-service'",
	}
	opts.MergeStrategy = ExportMergeStrategyReplaceEnvs
	staticEnv, err := FindEnvs(t.Context(), "test-export-envs", FindOpts{Selector: labels.SelectorFromSet(labels.Set{"type": "static"})})
	require.NoError(t, err)
	require.NoError(t, ExportEnvironments(t.Context(), staticEnv, tempDir, opts))
	checkFiles(t, tempDir, []string{
		filepath.Join(tempDir, "inline-namespace1", "my-configmap.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-service.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-service.yaml"),
		filepath.Join(tempDir, "static", "updated-deployment.yaml"),
		filepath.Join(tempDir, "static", "updated-service.yaml"),
		filepath.Join(tempDir, "manifest.json"),
	})
	manifestContent, err = os.ReadFile(filepath.Join(tempDir, "manifest.json"))
	require.NoError(t, err)
	assert.Equal(t, string(manifestContent), `{
    "inline-namespace1/my-configmap.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace1/my-deployment.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace1/my-service.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace2/my-deployment.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "inline-namespace2/my-service.yaml": "test-export-envs/inline-envs/main.jsonnet",
    "static/updated-deployment.yaml": "test-export-envs/static-env/main.jsonnet",
    "static/updated-service.yaml": "test-export-envs/static-env/main.jsonnet"
}`)

	// Re-export and delete the files of one env
	opts.Opts.ExtCode = jsonnet.InjectedCode{
		"deploymentName": "'updated-again-deployment'",
		"serviceName":    "'updated-again-service'",
	}
	opts.MergeDeletedEnvs = []string{"test-export-envs/inline-envs/main.jsonnet"}
	require.NoError(t, ExportEnvironments(t.Context(), staticEnv, tempDir, opts))
	checkFiles(t, tempDir, []string{
		filepath.Join(tempDir, "static", "updated-again-deployment.yaml"),
		filepath.Join(tempDir, "static", "updated-again-service.yaml"),
		filepath.Join(tempDir, "manifest.json"),
	})
	manifestContent, err = os.ReadFile(filepath.Join(tempDir, "manifest.json"))
	require.NoError(t, err)
	assert.Equal(t, string(manifestContent), `{
    "static/updated-again-deployment.yaml": "test-export-envs/static-env/main.jsonnet",
    "static/updated-again-service.yaml": "test-export-envs/static-env/main.jsonnet"
}`)
}

func TestExportEnvironmentsBroken(t *testing.T) {
	tempDir := t.TempDir()
	require.NoError(t, os.Chdir("testdata"))
	defer func() { require.NoError(t, os.Chdir("..")) }()

	// Find envs
	envs, err := FindEnvs(t.Context(), "test-export-envs-broken", FindOpts{Selector: labels.Everything()})
	require.NoError(t, err)

	// Export all envs
	opts := &ExportEnvOpts{
		Format:    "{{.metadata.namespace}}/{{.metadata.name}}",
		Extension: "yaml",
	}

	var schemaError *manifest.SchemaError
	require.ErrorAs(t, ExportEnvironments(t.Context(), envs, tempDir, opts), &schemaError)
}

func BenchmarkExportEnvironmentsWithReplaceEnvs(b *testing.B) {
	zerolog.SetGlobalLevel(zerolog.Disabled)
	tempDir := b.TempDir()
	require.NoError(b, os.Chdir("testdata"))
	defer func() { require.NoError(b, os.Chdir("..")) }()

	// Find envs
	envs, err := FindEnvs(b.Context(), "test-export-envs", FindOpts{Selector: labels.Everything()})
	require.NoError(b, err)

	// Export all envs
	opts := &ExportEnvOpts{
		Format:        "{{.metadata.namespace}}/{{.metadata.name}}",
		Extension:     "yaml",
		MergeStrategy: ExportMergeStrategyReplaceEnvs,
	}
	opts.Opts.ExtCode = jsonnet.InjectedCode{
		"deploymentName": "'initial-deployment'",
		"serviceName":    "'initial-service'",
	}
	// Export a first time so that the benchmark loops are identical
	require.NoError(b, ExportEnvironments(b.Context(), envs, tempDir, opts))

	// On every loop, delete manifests from previous envs + reexport all envs
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		require.NoError(b, ExportEnvironments(b.Context(), envs, tempDir, opts), "failed on iteration %d", i)
	}
}

func TestExportEnvironmentsSkipManifest(t *testing.T) {
	tempDir := t.TempDir()
	require.NoError(t, os.Chdir("testdata"))
	defer func() { require.NoError(t, os.Chdir("..")) }()

	// Find envs
	envs, err := FindEnvs(t.Context(), "test-export-envs", FindOpts{Selector: labels.Everything()})
	require.NoError(t, err)

	// Export all envs with skip manifest flag
	opts := &ExportEnvOpts{
		Format:       "{{.metadata.namespace}}/{{.metadata.name}}",
		Extension:    "yaml",
		SkipManifest: true,
	}
	opts.Opts.ExtCode = jsonnet.InjectedCode{
		"deploymentName": "'test-deployment'",
		"serviceName":    "'test-service'",
	}
	require.NoError(t, ExportEnvironments(t.Context(), envs, tempDir, opts))
	
	// Check that all manifest files are created but manifest.json is NOT created
	expectedFiles := []string{
		filepath.Join(tempDir, "inline-namespace1", "my-configmap.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace1", "my-service.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-deployment.yaml"),
		filepath.Join(tempDir, "inline-namespace2", "my-service.yaml"),
		filepath.Join(tempDir, "static", "test-deployment.yaml"),
		filepath.Join(tempDir, "static", "test-service.yaml"),
	}
	checkFiles(t, tempDir, expectedFiles)
	
	// Explicitly verify manifest.json does not exist
	manifestPath := filepath.Join(tempDir, "manifest.json")
	_, err = os.Stat(manifestPath)
	assert.True(t, os.IsNotExist(err), "manifest.json should not exist when SkipManifest is true")
}

func checkFiles(t testing.TB, dir string, files []string) {
	t.Helper()

	var existingFiles []string
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		existingFiles = append(existingFiles, path)
		return nil
	})
	require.NoError(t, err)

	assert.ElementsMatch(t, files, existingFiles)
}

File: rust-export/src/export_test.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::Result;
use walkdir::WalkDir;

use crate::export::{export_environments, ExportMergeStrategy, ExportOptions};

fn check_files(dir: &Path, expected_files: &[&str]) -> Result<()> {
    let mut existing_files = Vec::new();
    for entry in WalkDir::new(dir).into_iter().filter_map(|e| e.ok()) {
        if entry.file_type().is_file() {
            existing_files.push(entry.path().to_path_buf());
        }
    }

    let expected: Vec<PathBuf> = expected_files.iter().map(PathBuf::from).collect();

    // Sort for comparison
    let mut existing_sorted = existing_files.clone();
    existing_sorted.sort();
    let mut expected_sorted = expected.clone();
    expected_sorted.sort();

    if existing_sorted != expected_sorted {
        eprintln!("Expected files:");
        for f in &expected_sorted {
            eprintln!("  {:?}", f);
        }
        eprintln!("Actual files:");
        for f in &existing_sorted {
            eprintln!("  {:?}", f);
        }
        panic!("File lists don't match");
    }

    Ok(())
}

#[test]
fn test_export_environments() -> Result<()> {
    let temp_dir = tempfile::tempdir()?;
    let output_dir = temp_dir.path().to_path_buf();

    // Change to testdata directory
    let original_dir = std::env::current_dir()?;
    let testdata_dir = original_dir.join("testdata").canonicalize()?;
    std::env::set_current_dir(&testdata_dir)?;

    let result = (|| -> Result<()> {
        // Export all envs
        let paths = vec![PathBuf::from("test-export-envs")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::None,
            merge_deleted_envs: vec![],
            skip_manifest: false,
            targets: vec![],
            name: None,
            selector: None,
        };

        export_environments(output_dir.clone(), paths, true, opts)?;

        // Check exported files
        check_files(
            &output_dir,
            &[
                &format!(
                    "{}/inline-namespace1/my-configmap.yaml",
                    output_dir.display()
                ),
                &format!(
                    "{}/inline-namespace1/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace1/my-service.yaml", output_dir.display()),
                &format!(
                    "{}/inline-namespace2/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace2/my-service.yaml", output_dir.display()),
                &format!("{}/static/initial-deployment.yaml", output_dir.display()),
                &format!("{}/static/initial-service.yaml", output_dir.display()),
                &format!("{}/manifest.json", output_dir.display()),
            ],
        )?;

        // Check manifest.json content
        let manifest_path = output_dir.join("manifest.json");
        let manifest_content = fs::read_to_string(&manifest_path)?;
        let manifest: HashMap<String, String> = serde_json::from_str(&manifest_content)?;

        let mut expected_manifest = HashMap::new();
        expected_manifest.insert(
            "inline-namespace1/my-configmap.yaml".to_string(),
            "test-export-envs/inline-envs/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "inline-namespace1/my-deployment.yaml".to_string(),
            "test-export-envs/inline-envs/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "inline-namespace1/my-service.yaml".to_string(),
            "test-export-envs/inline-envs/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "inline-namespace2/my-deployment.yaml".to_string(),
            "test-export-envs/inline-envs/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "inline-namespace2/my-service.yaml".to_string(),
            "test-export-envs/inline-envs/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "static/initial-deployment.yaml".to_string(),
            "test-export-envs/static-env/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "static/initial-service.yaml".to_string(),
            "test-export-envs/static-env/main.jsonnet".to_string(),
        );

        assert_eq!(
            manifest, expected_manifest,
            "Manifest content doesn't match"
        );

        // Try to re-export - should fail
        let paths = vec![PathBuf::from("test-export-envs")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::None,
            merge_deleted_envs: vec![],
            skip_manifest: false,
            targets: vec![],
            name: None,
            selector: None,
        };

        let result = export_environments(output_dir.clone(), paths, true, opts);
        assert!(result.is_err(), "Should fail when directory is not empty");
        assert!(
            result.unwrap_err().to_string().contains("not empty"),
            "Error should mention directory not empty"
        );

        // Try with fail-on-conflicts
        let paths = vec![PathBuf::from("test-export-envs")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::FailOnConflicts,
            merge_deleted_envs: vec![],
            skip_manifest: false,
            targets: vec![],
            name: None,
            selector: None,
        };

        let result = export_environments(output_dir.clone(), paths, true, opts);
        assert!(result.is_err(), "Should fail on conflicts");
        assert!(
            result.unwrap_err().to_string().contains("already exists"),
            "Error should mention file already exists"
        );

        // Re-export only static env with replace-envs strategy
        let paths = vec![PathBuf::from("test-export-envs/static-env")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::ReplaceEnvs,
            merge_deleted_envs: vec![],
            skip_manifest: false,
            targets: vec![],
            name: None,
            selector: None,
        };

        export_environments(output_dir.clone(), paths, false, opts)?;

        check_files(
            &output_dir,
            &[
                &format!(
                    "{}/inline-namespace1/my-configmap.yaml",
                    output_dir.display()
                ),
                &format!(
                    "{}/inline-namespace1/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace1/my-service.yaml", output_dir.display()),
                &format!(
                    "{}/inline-namespace2/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace2/my-service.yaml", output_dir.display()),
                &format!("{}/static/initial-deployment.yaml", output_dir.display()),
                &format!("{}/static/initial-service.yaml", output_dir.display()),
                &format!("{}/manifest.json", output_dir.display()),
            ],
        )?;

        // Re-export and delete the files of inline envs
        let paths = vec![PathBuf::from("test-export-envs/static-env")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::ReplaceEnvs,
            merge_deleted_envs: vec!["test-export-envs/inline-envs/main.jsonnet".to_string()],
            skip_manifest: false,
            targets: vec![],
            name: None,
            selector: None,
        };

        export_environments(output_dir.clone(), paths, false, opts)?;

        check_files(
            &output_dir,
            &[
                &format!("{}/static/initial-deployment.yaml", output_dir.display()),
                &format!("{}/static/initial-service.yaml", output_dir.display()),
                &format!("{}/manifest.json", output_dir.display()),
            ],
        )?;

        let manifest_content = fs::read_to_string(&manifest_path)?;
        let manifest: HashMap<String, String> = serde_json::from_str(&manifest_content)?;

        let mut expected_manifest = HashMap::new();
        expected_manifest.insert(
            "static/initial-deployment.yaml".to_string(),
            "test-export-envs/static-env/main.jsonnet".to_string(),
        );
        expected_manifest.insert(
            "static/initial-service.yaml".to_string(),
            "test-export-envs/static-env/main.jsonnet".to_string(),
        );

        assert_eq!(
            manifest, expected_manifest,
            "Manifest content doesn't match after deletion"
        );

        Ok(())
    })();

    // Always restore the original directory
    std::env::set_current_dir(original_dir)?;

    result
}

#[test]
fn test_export_environments_skip_manifest() -> Result<()> {
    let temp_dir = tempfile::tempdir()?;
    let output_dir = temp_dir.path().to_path_buf();

    let original_dir = std::env::current_dir()?;
    let testdata_dir = original_dir.join("testdata").canonicalize()?;
    std::env::set_current_dir(&testdata_dir)?;

    let result = (|| -> Result<()> {
        let paths = vec![PathBuf::from("test-export-envs")];
        let opts = ExportOptions {
            format: "{{.metadata.namespace}}/{{.metadata.name}}".to_string(),
            extension: "yaml".to_string(),
            parallelism: 1,
            cache_path: None,
            cache_path_regexes: vec![],
            merge_strategy: ExportMergeStrategy::None,
            merge_deleted_envs: vec![],
            skip_manifest: true,
            targets: vec![],
            name: None,
            selector: None,
        };

        export_environments(output_dir.clone(), paths, true, opts)?;

        // Check that manifest files exist but manifest.json does NOT exist
        check_files(
            &output_dir,
            &[
                &format!(
                    "{}/inline-namespace1/my-configmap.yaml",
                    output_dir.display()
                ),
                &format!(
                    "{}/inline-namespace1/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace1/my-service.yaml", output_dir.display()),
                &format!(
                    "{}/inline-namespace2/my-deployment.yaml",
                    output_dir.display()
                ),
                &format!("{}/inline-namespace2/my-service.yaml", output_dir.display()),
                &format!("{}/static/initial-deployment.yaml", output_dir.display()),
                &format!("{}/static/initial-service.yaml", output_dir.display()),
            ],
        )?;

        // Verify manifest.json does not exist
        let manifest_path = output_dir.join("manifest.json");
        assert!(
            !manifest_path.exists(),
            "manifest.json should not exist when skip_manifest is true"
        );

        Ok(())
    })();

    std::env::set_current_dir(original_dir)?;

    result
}
